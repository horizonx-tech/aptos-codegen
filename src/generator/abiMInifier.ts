import {
  MoveModuleJSON,
  MoveModuleJSONMinified,
} from '@horizonx/aptos-module-client'
import { isEventHandle, isResource } from './utils'

type ABIMinifierOptions = {
  entryFunctionsDisabled?: boolean
  gettersDisabled?: boolean
}
export const minifyABI = (
  abi: MoveModuleJSON,
  options?: ABIMinifierOptions,
): MoveModuleJSONMinified => ({
  name: abi.name,
  address: abi.address,
  exposed_functions: options?.entryFunctionsDisabled
    ? []
    : abi.exposed_functions
        .filter(({ is_entry }) => is_entry)
        .map(({ name, is_entry, generic_type_params, params }) => ({
          name,
          is_entry,
          generic_type_params: generic_type_params.map(() => ({})),
          params: params.filter((param) => param !== '&signer'),
        })),
  structs: options?.gettersDisabled
    ? []
    : abi.structs.filter(isResource).map(({ name, abilities, fields }) => ({
        name,
        abilities: abilities.filter((ability) => ability === 'key'),
        fields: fields.map((field) =>
          isEventHandle(field) ? field : { name: field.name },
        ),
      })),
})
