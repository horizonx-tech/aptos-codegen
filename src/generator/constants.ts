import { MoveType } from 'aptos/dist/generated'

export const JS_NATIVE_TYPES = ['boolean', 'string', 'Array', 'any']

export const RESERVERD_MODULES = [
  { key: 'aptos', types: ['BCS', 'MaybeHexString', 'Types'] },
  {
    key: '@horizonx/aptos-module-client',
    types: [
      'AptosModuleClient',
      'TypedMoveResource',
      'TypedEvent',
      'EventGetterParams',
    ],
  },
]

export const RESERVED_TYPE_DICT: Partial<Record<MoveType, string>> = {
  bool: 'boolean',
  u8: 'BCS.Uint8',
  u16: 'BCS.Uint16',
  u32: 'BCS.Uint32',
  u64: 'BCS.Uint64',
  u128: 'BCS.Uint128',
  address: 'MaybeHexString',
  vector: 'Array',
  '0x1::string::String': 'string',
}

export const RESOURCE_TYPE_REGEX = /^(\w*::\w*::\w*)(?:<(.*)>)?/
export const TYPE_PARAMETER_REGEX = /^T[0-9]+$/
