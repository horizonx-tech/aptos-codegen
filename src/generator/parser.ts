import { MoveModuleJSON } from '@horizonx/aptos-module-client'
import { MoveType } from 'aptos/dist/generated'
import { pascalCase } from 'change-case'
import {
  EventHandleTypeStruct,
  EventStruct,
  FunctionStruct,
  ModuleStruct,
  ResourceStruct,
  StructStruct,
  TypeStruct,
} from '../types'
import {
  hasEventHandle,
  isEventHandle,
  isJsNativeType,
  isResource,
  isString,
  isTypeParameter,
  notSigner,
  parseFromResourceType,
  toReservedType,
} from './utils'

export const parseFromABI = (abi: MoveModuleJSON): ModuleStruct => {
  const id = `${abi.address}::${abi.name}`

  const entryFunctions = abi.exposed_functions
    .filter(({ is_entry }) => is_entry)
    .map(({ name, generic_type_params, params }) => ({
      name,
      typeArguments: generic_type_params,
      args: params.filter(notSigner).map(toTypeStruct),
    }))

  const resourceStructs = abi.structs.filter(isResource)

  const resources = resourceStructs.map(({ name }) => ({ name }))
  const events = resourceStructs.filter(hasEventHandle).flatMap(({ fields }) =>
    fields.filter(isEventHandle).map(({ name, type }) => ({
      name: pascalCase(name),
      type: toTypeStruct(type) as EventHandleTypeStruct,
    })),
  )

  const structs = abi.structs.map(({ name, fields }) => ({
    name,
    fields: fields.map(({ name, type }) => ({
      name,
      type: toTypeStruct(type),
    })),
  }))

  const dependencies = extractDependencies(
    id,
    entryFunctions,
    resources,
    events,
    structs,
  )

  return {
    id,
    address: abi.address,
    name: pascalCase(abi.name),
    abi,
    entryFunctions,
    resources,
    events,
    structs,
    dependencies,
  }
}

export const extractDependencies = (
  moduleId: string,
  entryFunctions: FunctionStruct[],
  resources: ResourceStruct[],
  events: EventStruct[],
  structs: StructStruct[],
) => {
  const allTypeStructs = [
    ...entryFunctions.flatMap(({ args }) => args),
    ...structs.flatMap(({ fields }) => fields.flatMap(({ type }) => type)),
  ]
  const additionalDependencies = [
    ...(resources.length ? ['address', 'TypedMoveResource', 'Types'] : []),
    ...(entryFunctions.length || resources.length ? ['AptosModuleClient'] : []),
    ...(events.length ? ['TypedEvent', 'EventGetterParams'] : []),
  ]
  return Array.from(
    new Set([
      ...allTypeStructs
        .flatMap(extractTypeNameRecursive)
        .filter(
          (type) =>
            !type.startsWith(moduleId) && !isJsNativeType(toReservedType(type)),
        ),
      ...additionalDependencies,
    ]),
  )
}

export const toTypeStruct = (param: MoveType): TypeStruct => {
  const reservedType = toReservedType(param)
  if (reservedType) return param
  if (isTypeParameter(param)) return param
  if (param.startsWith('vector')) {
    const genericTypes = extractIdentifiersFromGenericTypeStr(
      param.replace(/^.*?</, '').replace(/>$/, ''),
    )
    return {
      name: 'vector',
      genericTypes: genericTypes.map(toTypeStruct),
    }
  }
  const res = parseFromResourceType(param)
  if (!res) {
    console.warn(`Skip unknown type: ${param}`)
    return 'any'
  }
  return {
    moduleId: res.moduleId,
    name: res.structName,
    genericTypes: res.genericTypesStr
      ? extractIdentifiersFromGenericTypeStr(res.genericTypesStr).map(
          toTypeStruct,
        )
      : [],
  }
}

export const extractIdentifiersFromGenericTypeStr = (str: string) => {
  const result = []
  let current = ''
  for (const each of str.split(', ')) {
    if (current) {
      current = [current, each].join(', ')
      if (!each.includes('>')) continue
      const numOfLeft = current.match(/</g).length
      const numOfRight = current.match(/>/g).length
      if (numOfLeft !== numOfRight) continue
      result.push(current)
      current = ''
      continue
    }
    if (each.includes('<') && !each.includes('>')) {
      current += each
      continue
    }
    result.push(each)
  }
  if (current) throw new Error(`Failed to parse generic types: ${str}`)
  return result
}

export const extractTypeNameRecursive = (type: TypeStruct): string[] => {
  if (isString(type)) return isTypeParameter(type) ? [] : [type]
  return [
    type.moduleId ? `${type.moduleId}::${type.name}` : type.name,
    ...type.genericTypes.flatMap(extractTypeNameRecursive),
  ]
}
