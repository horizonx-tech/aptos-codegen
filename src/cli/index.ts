#!/usr/bin/env node

import { exit } from 'process'
import { execute } from 'src/generator'
import { configure } from './config'

const main = async () => {
  const config = await configure(process.argv.slice(2))
  return execute(config)
}

main().catch((e) => {
  console.error(e)
  exit(1)
})
