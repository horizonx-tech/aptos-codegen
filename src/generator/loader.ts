import { AptosClient } from 'aptos'
import { MoveModuleJSON } from '@horizonx/aptos-module-client'
import { readFile } from 'fs/promises'
import { glob } from 'glob'
import { ModuleStruct } from 'src/types'
import { parseFromABI } from './parser'
import { toModuleId } from './utils'

export class ModuleLoader {
  private touchedModules = new Set<string>()
  private moduleMap = new Map<string, ModuleStruct>()
  private localABIMap = new Map<string, MoveModuleJSON>()

  constructor(private client: AptosClient) {}

  loadModules = async (moduleIds: string[], abiFilePathPatterns?: string[]) => {
    if (abiFilePathPatterns) await this.loadABIFromLocal(abiFilePathPatterns)
    await Promise.all(moduleIds.map(this.loadModule))
    return Array.from(this.moduleMap.values())
  }

  private loadModule = async (id: string) => {
    if (this.touchedModules.has(id)) return
    this.touchedModules.add(id)

    const abi = await this.loadABI(id)
    const module = parseFromABI(abi)
    this.moduleMap.set(module.id, module)

    await Promise.all(
      module.dependencies.map((dependency) => {
        const moduleId = toModuleId(dependency)
        return moduleId && this.loadModule(moduleId)
      }),
    )
  }

  private loadABI = async (id: string) => {
    const localABI = this.localABIMap.get(id)
    if (localABI) {
      console.log('loading module from local...', id)
      return localABI
    }
    console.log('loading module from chain...', id)
    return this.loadABIFromChain(id)
  }

  private loadABIFromLocal = async (patterns: string[]) => {
    const abis = await Promise.all(
      patterns.map(
        (pattern) =>
          new Promise<MoveModuleJSON[]>((resolve, reject) => {
            glob(pattern, {}, async (err, filePaths) => {
              if (err) reject(err)
              const res = await Promise.all(
                filePaths.map(async (filePath) => {
                  const json = JSON.parse(await readFile(filePath, 'utf8')) as
                    | MoveModuleJSON
                    | { abi: MoveModuleJSON }
                    | { abi: MoveModuleJSON }[]
                  if (Array.isArray(json)) return json.map(({ abi }) => abi)
                  if ('abi' in json) return [json.abi]
                  return [json]
                }),
              )
              return resolve(res.flat())
            })
          }),
      ),
    )
    abis.flat().forEach((abi) => {
      this.localABIMap.set(`${abi.address}::${abi.name}`, abi)
    })
  }

  private loadABIFromChain = async (id: string) => {
    const [address, name] = id.split('::')
    if (!address || !name) throw new Error(`Invalid moduleId: ${id}`)

    const { abi } = await this.client.getAccountModule(address, name)
    if (!abi) throw new Error(`ABI not found. moduleId: ${id}`)
    return abi
  }
}
