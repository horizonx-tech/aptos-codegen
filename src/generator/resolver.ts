import { StructDefinition, StructStruct, TypeStruct } from 'src/types'
import { RESERVERD_MODULES } from './constants'
import {
  hasTypeParameters as hasTypeParametersRecursive,
  isJsNativeType,
  isString,
  isTypeParameter,
  toModuleMapKey,
} from './utils'

export interface IModuleResolver {
  getDependenciesMap: () => Map<string, Set<string>>
  getStructDefinition: (moduleId: string, name: string) => StructDefinition
}

export class ModuleResolverFactory {
  private reservedModuleKeyMap = new Map<string, string>()
  private structMap = new Map<string, StructStruct>()
  private structDefinitionMap = new Map<string, StructDefinition>()

  constructor(
    modules: { id: string; name: string; structs: StructStruct[] }[],
  ) {
    this.resolveGenericTypeParameters(modules)
    RESERVERD_MODULES.forEach(({ key, types }) =>
      types.forEach((type) => this.reservedModuleKeyMap.set(type, key)),
    )
  }

  build = (dependencies: string[]): IModuleResolver => {
    const dependenciesMap = new Map<string, Set<string>>()
    for (const dependency of dependencies) {
      const moduleMapKey = toModuleMapKey(dependency)
      if (isJsNativeType(moduleMapKey)) continue
      const [address, moduleName, typeName] = dependency.split('::')
      const key = moduleName
        ? `${address}::${moduleName}`
        : this.reservedModuleKeyMap.get(moduleMapKey)
      if (!key) {
        console.warn('Skip: module id or reserved key not found:', dependency)
        continue
      }
      const type = typeName || moduleMapKey
      if (dependenciesMap.has(key)) dependenciesMap.get(key).add(type)
      else dependenciesMap.set(key, new Set([type]))
    }
    return new ModuleResolver(dependenciesMap, this.structDefinitionMap)
  }

  private resolveGenericTypeParameters = (
    modules: { id: string; structs: StructStruct[] }[],
  ) => {
    modules.forEach(({ id, structs }) =>
      structs.forEach((struct) =>
        this.structMap.set(`${id}::${struct.name}`, struct),
      ),
    )
    modules.forEach(({ id, structs }) => {
      structs.forEach((struct) =>
        this.resolveStructGenericTypeParametersRecursive(id, struct),
      )
    })
  }

  private resolveStructGenericTypeParametersRecursive = (
    moduleId: string,
    struct: StructStruct,
  ) => {
    const typeParameters = Array.from(
      new Set(
        struct.fields.flatMap(({ type }) =>
          this.resolveGenericTypeParametersRecursive(type),
        ),
      ),
    ).sort()
    this.structDefinitionMap.set(`${moduleId}::${struct.name}`, {
      typeParameters,
    })
    return typeParameters
  }

  private resolveGenericTypeParametersRecursive = (
    type: TypeStruct,
  ): string[] => {
    if (!isString(type)) {
      const genericTypes = type.genericTypes.filter(hasTypeParametersRecursive)
      if (!genericTypes.length) return []
    }
    return this.extractGenericType(type)
  }

  private extractGenericType = (type: TypeStruct): string[] => {
    if (isString(type)) return isTypeParameter(type) ? [type] : []

    if (!type.moduleId)
      return type.genericTypes.flatMap(this.extractGenericType)

    const fqn = `${type.moduleId}::${type.name}`
    const definition = this.structDefinitionMap.get(fqn)
    if (definition)
      return type.genericTypes
        .flatMap(this.resolveGenericTypeParametersRecursive)
        .slice(0, definition.typeParameters.length)

    const struct = this.structMap.get(fqn)
    if (!struct) {
      console.warn('Skip: unknown struct:', fqn)
      return []
    }
    return this.resolveStructGenericTypeParametersRecursive(
      type.moduleId,
      struct,
    )
  }
}

class ModuleResolver implements IModuleResolver {
  constructor(
    private dependenciesMap: Map<string, Set<string>>,
    private structDefinitionMap: Map<string, StructDefinition>,
  ) {}

  getDependenciesMap: IModuleResolver['getDependenciesMap'] = () =>
    this.dependenciesMap

  getStructDefinition: IModuleResolver['getStructDefinition'] = (
    moduleId,
    name,
  ) => this.structDefinitionMap.get(`${moduleId}::${name}`)
}
