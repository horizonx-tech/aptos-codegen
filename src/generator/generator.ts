import { ModuleStruct, TypeStruct } from 'src/types'
import { IModuleResolver } from './resolver'
import {
  entryFunction,
  eventsGetter,
  factory,
  imports,
  resourceGetter,
  resourceTypeGuard,
  struct,
  structField,
  toAlias,
  typeParametersExtaractor,
  types,
  typesContent,
  utilities,
} from './templates'
import {
  factoryFileName,
  isEventHandleFieldStruct,
  isResource,
  isTypeParameter,
  toPath,
  toReservedType,
  typesFileName,
  utilitiesFileName,
} from './utils'

type GeneratorParams = {
  module: ModuleStruct
  resolver: IModuleResolver
  factoryDisabled?: boolean
  alias?: string
}

export const generate = (params: GeneratorParams) => {
  const prefix = params.alias ? `./${params.alias}/` : undefined
  return {
    factory:
      params.factoryDisabled ||
      (params.module.entryFunctions.length === 0 &&
        params.module.resources.length === 0)
        ? undefined
        : {
            content: factory({
              name: params.module.name,
              abi: params.module.abi,
            }),
            path: toPath(factoryFileName(params.module.name), prefix),
          },
    types: {
      content: generateTypes(params),
      path: toPath(typesFileName(params.module.name), prefix),
    },
    utilities:
      params.module.resources.length === 0
        ? undefined
        : {
            content: generateUtilities(params),
            path: toPath(utilitiesFileName(params.module.name), prefix),
          },
  }
}

const generateUtilities = ({ module, resolver }: GeneratorParams) => {
  const resourceTypeGuards = module.resources.map(({ name }) => {
    const structDef = resolver.getStructDefinition(module.id, name)
    return resourceTypeGuard({
      moduleId: module.id,
      moduleName: module.abi.name,
      name,
      typeParameters: structDef.typeParameters,
    })
  })
  const typeParametersExtaractors = module.resources.map(({ name }) =>
    typeParametersExtaractor({
      moduleId: module.id,
      moduleName: module.abi.name,
      name,
    }),
  )
  return utilities({
    address: module.address,
    moduleName: module.name,
    resourceNames: module.resources.map(({ name }) => name),
    utilitiesContents: [...resourceTypeGuards, ...typeParametersExtaractors],
  })
}

const generateTypes = (params: GeneratorParams) => {
  const dependenciesMap = params.resolver.getDependenciesMap()
  const typeCount = Array.from(dependenciesMap.values())
    .flatMap((values) => Array.from(values.types))
    .reduce<Record<string, number>>(
      (res, type) => ({
        ...res,
        [type]: (res[type] || 0) + 1,
      }),
      {},
    )
  const importsContent = imports(
    params.module.id,
    dependenciesMap,
    typeCount,
    !!params.alias,
  )
  return types({
    importsContent,
    typesContent: generateTypesContent(params, typeCount),
  })
}

const generateTypesContent = (
  { module, resolver }: GeneratorParams,
  typeCount: Record<string, number>,
) => {
  const entryFunctions = module.entryFunctions.map(
    ({ name, typeArguments, args }) =>
      entryFunction({
        name,
        typeArguments,
        args: args.map((type) => toTSType(type, resolver)),
      }),
  )

  const resourceGetters = module.resources.map(({ name }) => {
    const structDef = resolver.getStructDefinition(module.id, name)
    return resourceGetter({
      name,
      typeParameters: structDef.typeParameters,
    })
  })

  const eventsGetters = module.structs
    .filter(isResource)
    .flatMap(({ fields }) =>
      fields.filter(isEventHandleFieldStruct).map(({ name, type }) => {
        const genericType = type.genericTypes[0]
        const structDef = resolver.getStructDefinition(
          genericType.moduleId,
          genericType.name,
        )
        return eventsGetter({
          name,
          type: toTSType(genericType, resolver),
          typeParameters: structDef.typeParameters,
        })
      }),
    )

  const structs = module.structs.map((each) => {
    const fields = each.fields.map(({ name, type }) =>
      structField({
        name: name,
        type: toTSType(type, resolver, typeCount),
      }),
    )
    const structDef = resolver.getStructDefinition(module.id, each.name)
    return struct({
      name: each.name,
      fields,
      typeParameters: structDef.typeParameters,
    })
  })

  return typesContent({
    name: module.name,
    entryFunctions,
    resourceGetters,
    eventsGetters,
    structs,
  })
}

const toTSType = (
  { moduleId, name, genericTypes }: TypeStruct,
  resolver: IModuleResolver,
  typeCount: Record<string, number> = {},
): string => {
  if (isTypeParameter(name)) return name
  const reservedType = toReservedType(name, moduleId)
  if (reservedType) {
    if (!genericTypes?.length) return reservedType
    const genericTypeNames = genericTypes.map((type) =>
      toTSType(type, resolver, typeCount),
    )
    return `${reservedType}<${genericTypeNames.join(', ')}>`
  }
  const structDef = resolver.getStructDefinition(moduleId, name)
  if (!structDef) return 'any'
  const typeName = typeCount[name] > 1 ? toAlias(moduleId, name) : name
  if (!structDef.typeParameters?.length) return typeName
  return `${typeName}<${genericTypes
    .slice(0, structDef.typeParameters.length)
    .map((each) => toTSType(each, resolver, typeCount))
    .join(', ')}>`
}
