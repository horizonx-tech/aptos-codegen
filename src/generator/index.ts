import { AptosClient } from 'aptos'
import { Config } from 'src/types'
import { generate } from './generator'
import { ModuleLoader } from './loader'
import { ModuleResolverFactory } from './resolver'
import { writeFiles } from './writer'

export const execute = async (config: Config) => {
  const files = await generateFiles(config)
  return writeFiles(files, config.outDir)
}

export const generateFiles = async (config: Config) => {
  const { modules, aliases } = await new ModuleLoader(
    new AptosClient(config.nodeUrl),
  ).loadModules(config.modules, config)
  const resolverFactory = new ModuleResolverFactory(modules, aliases)
  return modules.map((module) =>
    generate({
      module,
      resolver: resolverFactory.build(module.id, module.dependencies),
      factoryDisabled: !config.modules.includes(module.id),
      alias: aliases[module.address],
    }),
  )
}
