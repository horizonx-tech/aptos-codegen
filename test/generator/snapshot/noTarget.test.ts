import { generateFromModules } from 'src/generator'

it('snapshot no target', async () => {
  const files = await generateFromModules({
    nodeUrl: 'https://fullnode.devnet.aptoslabs.com/v1',
    outDir: 'test/__snapshots__',
    modules: ['0x1::coin'],
    abiFilePathPatterns: ['abi/**/*.json'],
    minifyAbi: true,
    targets: {},
  })
  expect(
    files.sort((a, b) => (a.types.path > b.types.path ? 1 : -1)),
  ).toMatchSnapshot()
})
