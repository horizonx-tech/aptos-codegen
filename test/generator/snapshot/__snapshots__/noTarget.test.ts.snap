// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot no target 1`] = `
[
  {
    "factory": undefined,
    "types": {
      "content": "/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { BCS, MaybeHexString } from 'aptos'

export type Aggregator = {
  handle: MaybeHexString
  key: MaybeHexString
  limit: BCS.Uint128
}
",
      "path": "./Aggregator.ts",
    },
    "utilities": undefined,
  },
  {
    "factory": undefined,
    "types": {
      "content": "/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { BCS, MaybeHexString, Types } from 'aptos'
import { Option } from './Option'
import { OptionalAggregator } from './OptionalAggregator'
import { EventHandle } from './Event'
import { AptosModuleClient, EventGetterParams, TypedMoveResource, TypedVersionedEvent } from '@horizonx/aptos-module-client'

export type BurnCapability = {
  dummy_field: boolean
}

export type Coin = {
  value: BCS.Uint64
}

export type CoinInfo = {
  name: string
  symbol: string
  decimals: BCS.Uint8
  supply: Option<OptionalAggregator>
}

export type CoinStore = {
  coin: Coin
  frozen: boolean
  deposit_events: EventHandle
  withdraw_events: EventHandle
}

export type DepositEvent = {
  amount: BCS.Uint64
}

export type FreezeCapability = {
  dummy_field: boolean
}

export type MintCapability = {
  dummy_field: boolean
}

export type SupplyConfig = {
  allow_upgrades: boolean
}

export type WithdrawEvent = {
  amount: BCS.Uint64
}
",
      "path": "./Coin.ts",
    },
    "utilities": undefined,
  },
  {
    "factory": undefined,
    "types": {
      "content": "/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { BCS } from 'aptos'
import { GUID } from './Guid'

export type EventHandle = {
  counter: BCS.Uint64
  guid: GUID
}
",
      "path": "./Event.ts",
    },
    "utilities": undefined,
  },
  {
    "factory": undefined,
    "types": {
      "content": "/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { BCS, MaybeHexString } from 'aptos'

export type GUID = {
  id: ID
}

export type ID = {
  creation_num: BCS.Uint64
  addr: MaybeHexString
}
",
      "path": "./Guid.ts",
    },
    "utilities": undefined,
  },
  {
    "factory": undefined,
    "types": {
      "content": "/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
export type Option<T0> = {
  vec: Array<T0>
}
",
      "path": "./Option.ts",
    },
    "utilities": undefined,
  },
  {
    "factory": undefined,
    "types": {
      "content": "/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { BCS } from 'aptos'
import { Option } from './Option'
import { Aggregator } from './Aggregator'

export type Integer = {
  value: BCS.Uint128
  limit: BCS.Uint128
}

export type OptionalAggregator = {
  aggregator: Option<Aggregator>
  integer: Option<Integer>
}
",
      "path": "./OptionalAggregator.ts",
    },
    "utilities": undefined,
  },
]
`;
